### DISCUSSION TITLE
Next.js deployment in Plesk panel occurs page reloading issues

### DISCUSSION THREAD

**Discussion by mohaimenmahi** [vercel/next.js#19623] [Help]
I am using Plesk Panel for deploying my nextjs application. Let's assume my domain is **example.com**. The home page redirects to **example.com/main**. I can go to any page by clicking and navigating on the page and the URL becomes **example.com/profile** or wherever I go. But when I reload the page or I hit the URL **example.com/profile** from another tab, it redirects to the **example.com/main** page. This behavior is happening for any internal pages I am visiting. I am using DigitalOcean with centOs in the Plesk panel for hosting. How can I solve this? 

N.B: Meanwhile I also deployed a version of my application on Vercel and that works fine. But I have to use Plesk with Digitalocean and CentOS integrated hosting for some reasons.

**Comment by mohaimenmahi** [comment]
solved by #10522

**Comment by brianwachira** [comment]
@mohaimenmahi I am currently trying to deploy my nextjs app on a Plesk server. Do you have a step by step tutorial on how I can achieve this?

**Reply by mohaimenmahi** [reply]
@brianwachira I just edited the `.htaccess` file. Here I am providing the .htacces for root file:

```
<IfModule mod_rewrite.c>
RewriteEngine On 
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME}\.html -f
RewriteRule ^(.*)$ $1.html [NC,L]
</IfModule>
```

for nested pages, the add this .htaccess file to the folder which contains pages:

```
<IfModule mod_rewrite.c>
  RewriteEngine On
  RewriteBase /product
  RewriteRule ^[slug]\.html$ - [L]
  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteCond %{REQUEST_FILENAME} !-l
  RewriteRule . /product/[slug].html [L]
</IfModule>
```
This is under my `/pages/product` folder and I have a `[slug].tsx` file here for dynamic routing.

**Reply by brianwachira** [reply]
Alright. 
This is how ive set up nodejs on plesk
![Screenshot 2021-06-18 at 09-16-34 Node js on mymoticarshow com - Eupotech Limited](https://user-images.githubusercontent.com/28436092/122515412-15e32e00-d016-11eb-9cd0-bd78ed8cc109.png)

This is not working. The site shows : 

> Incomplete response received from application

Could you maybe have a clue on how I can set up my nextjs well on plesk @mohaimenmahi

**Reply by brianwachira** [reply]
> Alright.
> This is how ive set up nodejs on plesk
> ![Screenshot 2021-06-18 at 09-16-34 Node js on mymoticarshow com - Eupotech Limited](https://user-images.githubusercontent.com/28436092/122515412-15e32e00-d016-11eb-9cd0-bd78ed8cc109.png)
> 
> This is not working. The site shows :
> 
> > Incomplete response received from application
> 
> Could you maybe have a clue on how I can set up my nextjs well on plesk @mohaimenmahi

*Update*  - I changed the set up but am still getting the same message on my webpage
![Screenshot 2021-06-18 at 11-08-03 Node js on mymoticarshow com - Eupotech Limited](https://user-images.githubusercontent.com/28436092/122529139-d02e6180-d025-11eb-8a72-15676bffdc2d.png)

**Reply by mohaimenmahi** [reply]
You do not need node setup or start this in node environment. Update the script on your `package.json` file like this:

```
"scripts": {
    "dev": "next",
    "build": "next build && next export -o output",
    "start": "next start -p $PORT",
    "shop-dev": "next"
  }
```
Then run `yarn build` on your terminal.
This will create an `output` folder on your project directory.

Then upload the files you have the files on your output folder to your domain folder. add .htaccess file. Thats it.

**Reply by mohaimenmahi** [reply]
Just forget about the node environment and running it.

**Reply by brianwachira** [reply]
It has worked. Thank you very much :pray:

**Reply by brianwachira** [reply]
Hi there, let me just explain what I was trying to do.
I built a nextjs app that consumes from strapi  headlesscms . The next export method works and the site is visible.
However when I make changes on the headless cms, they don't reflect on the site, yet I'm fetching data using getstaticprops. I even added revalidate. After much digging I read that statically exporting an application via 'next export' disables api routes. And that after build you have to run next start so that revalidate works. 

I wanted to ask, do you know how I can host next app on plesk server without having to do next export and in a way that revalidate works?

**Reply by mohaimenmahi** [reply]
Are you using Apollo Client for GraphQL?

**Reply by brianwachira** [reply]
No am not using any graphql. I am fetching my data using fetch

**Reply by mohaimenmahi** [reply]
You can try `getServerSideProps` instead of `getStaticProps`.

**Reply by brianwachira** [reply]
Okay, what about the performance?

**Reply by mohaimenmahi** [reply]
Not varies that much

**Reply by brianwachira** [reply]
When I used `getServerSideProps`, I got a next export error. Turns out you can't next export  while using getServerSideProps.
I ended up running next with a custom server on Plesk. 
here's my custom server code :-
//app.js
`const { `
 `createServer`
`} = require("http");`
`const {`
 ` parse`
`} = require("url");`
`const next = require("next");`

`const port = process.env.PORT || 3000;`

`// Create the Express-Next App`
`const app = next({`
 ` dev:false`
`});`
`const handle = app.getRequestHandler();`

`app`
 ` .prepare()`
  `.then(() => {`
   ` createServer((req, res) => {`
    `  const parsedUrl = parse(req.url, true);`
     ` const {`
      `  pathname,`
       ` query`
      `} = parsedUrl;`
     ` handle(req, res, parsedUrl);`
      `console.log("pathname", pathname);`
    `}).listen(port, (err) => {`
     ` if (err) throw err;`
      `console.log('Ready on http://localhost:${port}`);
`    });`
`  })`
`  .catch((ex) => {`
 `   console.error(ex.stack);`
  `  process.exit(1);`
 ` });`
I then pointed plesk document root to the build folder i.e `.next ` folder, then the application startup file to `app.js`
